lib/AST/Module.cpp

```diff
@@ -2662,6 +2664,70 @@ SourceFile::lookupOpaqueResultType(StringRef MangledName) {
   return nullptr;
 }
 
+//===----------------------------------------------------------------------===//
+// SynthesizedFileUnit Implementation
+//===----------------------------------------------------------------------===//
+
+SynthesizedFileUnit::SynthesizedFileUnit(ModuleDecl &M)
+    : FileUnit(FileUnitKind::Synthesized, M) {
+  M.getASTContext().addDestructorCleanup(*this);
+}
+
+Identifier
+SynthesizedFileUnit::getDiscriminatorForPrivateValue(const ValueDecl *D) const {
+  assert(D->getDeclContext()->getModuleScopeContext() == this);
+
+  // Use cached primitive discriminator if it exists.
+  if (!PrivateDiscriminator.empty())
+    return PrivateDiscriminator;
+
+  assert(1 == count_if(getParentModule()->getFiles(),
+                       [](const FileUnit *FU) -> bool {
+                         return isa<SynthesizedFileUnit>(FU);
+                       }) &&
+         "Cannot promise uniqueness if multiple synthesized file units exist");
+
+  // Use a discriminator invariant across Swift version: a hash of the module
+  // name and a special string.
+  llvm::MD5 hash;
+  hash.update(getParentModule()->getName().str());
+  // TODO: Use a more robust discriminator for synthesized files. Pick something
+  // that cannot conflict with `SourceFile` discriminators.
+  hash.update("SYNTHESIZED FILE");
+  llvm::MD5::MD5Result result;
+  hash.final(result);
+
+  // Use the hash as a hex string, prefixed with an underscore to make sure
+  // it is a valid identifier.
+  // FIXME: There are more compact ways to encode a 16-byte value.
+  SmallString<33> buffer{"_"};
+  SmallString<32> hashString;
+  llvm::MD5::stringifyResult(result, hashString);
+  buffer += hashString;
+  PrivateDiscriminator = getASTContext().getIdentifier(buffer.str().upper());
+  return PrivateDiscriminator;
+}
+
+void SynthesizedFileUnit::lookupValue(
+    DeclName name, NLKind lookupKind,
+    SmallVectorImpl<ValueDecl *> &result) const {
+  for (auto *decl : TopLevelDecls) {
+    if (decl->getFullName().matchesRef(name))
+      result.push_back(decl);
+  }
```

Comment:
Note: `SourceFile` and `ModuleDecl` use `SourceLookupCache` to store declarations, which enables more efficient lookup. But I didn't use `SourceLookupCache` for `SynthesizedFileUnit` because that involves [publicly exposing a mutating `SourceLookupCache::addTopLevelDecl` function](https://github.com/apple/swift/pull/30845/files#diff-d6609c12b803cdc8cae7bd6b39a5db22R330-R332).

`TopLevelDecls` is a `SmallVector` instead of a `DenseMap`, copying the old [`DerivedFileUnit`](https://github.com/apple/swift/commit/211c58015878614f4cdebfd859fc3d1ba54f2fbd#diff-f5e0779c65aaf3ea695beda2f5acd7acL756). `DenseMap` makes lookup more efficient, but I think it makes `SynthesizedFileUnit::getTopLevelDecls` less efficient.